File: 104MaximumDepthOfBinaryTree.cpp
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {

public:
    int maxDepth(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        return max(maxDepth(root->left)+1, maxDepth(root->right)+1);
    }
};

int main() {
    Solution s;

    TreeNode* root = new TreeNode;
    TreeNode* node2 = new TreeNode(9);
    TreeNode* node3 = new TreeNode(20);
    TreeNode* node1 = new TreeNode(3, node2, node3);
    int depth = s.maxDepth(root);
    cout << depth;
}

File: 104MaximumDepthOfBinaryTree.py
from collections import deque

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # method1 - recursive dfs
    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        
        return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))
    
    # method2 - iterative bfs
    def iterativeBfs(self, root):
        if not root:
            return 0

        level = 0
        q = deque([root])

        while q:
            for i in range(len(q)):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            level += 1
        return level


    # method3 - iterative dfs
    def iterativeDfs(self, root):
        stack = [[root, 1]]
        result = 0

        while stack:
            node, depth = stack.pop()
            if node:
                result = max(result, depth)
                stack.append([node.left, depth+1])
                stack.append([node.right, depth+1])
        return result

if __name__ == "__main__":
    s = Solution()
    node15 = TreeNode(15)
    node7 = TreeNode(7)
    node9 = TreeNode(9)
    node20 = TreeNode(20, node15, node7)
    node3 = TreeNode(3, node9, node20)
    


